### ISSUE TITLE
Train on coco

### ISSUE THREAD

**Issue by ThanosRoidis** [xuannianz/EfficientDet#136]
I am trying to retrain EfficientDet-d0 on **coco** from scratch by following the instructions in the readme file:

- STEP1:
`python3 train.py --snapshot imagenet --phi {0, 1, 2, 3, 4, 5, 6} --gpu 0 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 1000 pascal|coco datasets/VOC2012|datasets/coco` to start training. The init lr is 1e-3.
- STEP2: 
`python3 train.py --snapshot xxx.h5 --phi {0, 1, 2, 3, 4, 5, 6} --gpu 0 --random-transform --compute-val-loss --freeze-bn --batch-size 4 --steps 10000 pascal|coco datasets/VOC2012|datasets/coco `to start training when val mAP can not increase during STEP1. The init lr is 1e-4 and decays to 1e-5 when val mAP keeps dropping down.

So far I have only run the first step with the following command:
`python3 train.py --weighted-bifpn --snapshot imagenet --phi 0--gpu 0 --random-transform --compute-val-loss --freeze-backbone --batch-size 32 --steps 1000 coco datasets/coco`
 (notice I have enabled `weighted-bifpn`)

Then I evaluated the last checkpoint on val2017 and got the following results:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.019
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.046
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.010
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.012
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.018
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.033
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.032
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.055
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.061
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.022
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.070
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.093
```

I know I should also run STEP2, but the results look quite bad as is. Should I change something in the hyperparameters of STEP1 / STEP2 or is this the expected behaviour? 

Also, about the comments regarding the learning rate:
> The init lr is 1e-3. (STEP1)
> The init lr is 1e-4 and decays to 1e-5 when val mAP keeps dropping down. (STEP2)

I did not notice any learning rate policies in the training code / callbacks, except in the optimizer:

```
 model.compile(optimizer=Adam(lr=1e-3), loss={
        'regression': smooth_l1_quad() if args.detect_quadrangle else smooth_l1(),
        'classification': focal()
    }, )
```

Is setting the lr something I should do manually in the code?

**Comment by charles-viss** [comment]
I am also attempting to train from scratch on coco and have run into a few issues.

The first is that there is an indexing discrepancy between the training and evaluation category labels. The labels defined in generators/coco.py only use the domain 0-79, while the labels used for evaluation in eval/coco.py are based on the original coco category ids (in the domain 1-90). I made the following change in generators/coco.py to address this:

annotations['labels'] = np.concatenate(
                #[annotations['labels'], [self.coco_label_to_label(a['category_id'])]], axis=0)
                [annotations['labels'], [int(a['category_id']) - 1]], axis=0)

With this I am able to reach an mAP of about 0.20 in STEP1. However, upon switch to STEP2 of training, the loss is dramatically increased from where it was at the end of STEP1, as if some of the model is not being loaded correctly from the snapshot.

And yes the lr is something you have to change manually in the code (or you can make an argument for it).

**Comment by SiBensberg** [comment]
I too have some problems with the training. Are you using the latest code? For me the latest code does not work either. The training mAP is starting very high at around ~0.3. When trying to inference some pictures after 30 Epochs just gibberish comes out (Just boxes with 0.010005 confidence and class 0)

**Comment by akmalulkhairin** [comment]
Hi, I am facing the same problem, have you managed to sort it out?

**Comment by ThanosRoidis** [comment]
> Hi, I am facing the same problem, have you managed to sort it out?

Nope, I gave up.

**Comment by charles-viss** [comment]
I did also notice that with the new model, the freeze-bn argument is no longer actually freezing the batch norm layers. I froze the batch norm layers manually and this prevented the loss from exploding when switching from step 1 to step 2, but I am still unable to get above an mAP of 0.23 with efficientdetd0

**Comment by nickkimer** [comment]
> I did also notice that with the new model, the freeze-bn argument is no longer actually freezing the batch norm layers. I froze the batch norm layers manually and this prevented the loss from exploding when switching from step 1 to step 2, but I am still unable to get above an mAP of 0.23 with efficientdetd0

how did you accomplish this? Switching from step 1 to step 2 gives me Map of ~ 0.6 to ~0.0013
